### PR TITLE
Telemetry: At Mentions

### PR DESCRIPTION
**Pull request by RXminuS**
Before the `cody.at-mention` event fired as soon as the chat view was rendered leading to an overcount of DAUs. Duplicate mention events were also fired for other mention providers such as when typing a file name where each character would fire an event.

Now events only fire when actually triggered. We also clean up and add additional public metadata that can help evaluate the actual effectiveness of different mention providers.

Fixes CODY-3405

## Test plan
- Added new Telemetry testing framework. Run `cd vscode & pnpm run test:e2e2 --grep "telemetry`.
- See the snapshots to view the included data

### CODE REVIEW

**Comment by RXminuS** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5558?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#5558** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5558?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#5026** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5026?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5278** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5278?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5120** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5120?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#4993](https://github.com/sourcegraph/cody/pull/4993) <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/4993?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#5228** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5228?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RXminuS and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/shared/src/prompt/truncation.ts [hunk 1]
```diff
@@ -1,17 +1,31 @@
...
+    text: PromptString,
+    maxTokens: number,
+    utils?: never
+): Promise<PromptString>
+export function truncatePromptString(text: PromptString, maxTokens: number, utils?: TokenCounterUtils) {
+    const exec = (rutils: TokenCounterUtils) => {
+        const encoded = rutils.encode(text.toString())
+        return encoded.length <= maxTokens
+            ? text
+            : text.slice(0, rutils.decode(encoded.slice(0, maxTokens))?.length).trim()
```
Comment (akalia25) [inline lib/shared/src/prompt/truncation.ts:21]: Making this function sync again would undo the changes in this [PR](https://github.com/sourcegraph/cody/pull/5231), would just like to be mindful that this is intended.

File: vscode/src/chat/chat-view/ChatController.ts [hunk 2]
```diff
@@ -811,12 +819,16 @@ export class ChatController implements vscode.Disposable, vscode.WebviewViewProv
...
+                    telemetryEvents['cody.chat-question/executed'].record(
+                        {
+                            ...telemetryProperties,
+                            context,
+                        },
+                        {
+                            addMetadata: true,
+                            current: span,
+                            firstToken: firstTokenSpan,
+                        }
```
Comment (akalia25) [inline vscode/src/chat/chat-view/ChatController.ts:831]: I'm a bit worried about this change, in terms of still capturing the same data. On`cody.chat-question/executed` I'm not seeing `promptText` appear in `privateMetadata` and `recordsPrivateMetadataTranscript` on `metadata`.

In addition, could we add the `isDotCom` check before recoding/storing the `promptText` or `ResponseText`, I'm not seeing that take place in the ChatController file anymore. 🤔

**Review by akalia25 — COMMENTED** [review]
Left a few clarifying comments :)

File: vscode/e2e/utils/vscody/fixture/telemetry.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                })
+                .intercept((req, res) => {
+                    const now = new Date()
+                    const body = req.jsonBody()
+                    res.status(200).json({
+                        data: { telemetry: { recordEvents: { alwaysNil: null } } },
+                    })
+                    const rawEvents = body?.variables?.events as any[]
+                    //todo: allow automatic failure of events
+                    for (const event of rawEvents) {
```
Comment (akalia25) [inline vscode/e2e/utils/vscody/fixture/telemetry.ts:47]: Would this be the right place to add the `logTestingData` [func](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/cody/-/blob/vscode/test/fixtures/mock-server.ts?L595) to send these events to our dedicated testing pubsub topic, that stores events from our E2E tests? Primarily curious, nothing to add to this PR :)

Comment (RXminuS) [reply vscode/e2e/utils/vscody/fixture/telemetry.ts:47]: I think that's probably best done after the test has executed similar to how the GithubCI and BuildKite reporters work

File: vscode/src/chat/chat-view/ChatController.ts (hunk 2 above)
Comment (RXminuS) [reply vscode/src/chat/chat-view/ChatController.ts:831]: There's an issue where files that are git-ignored but still added to context (such as `.env` files) are also stored as plaintext.

I spoke with @janhartman about it and since the data isn't used at all, and this will move to the backend, it felt unnecessary to come up with a complicated solution now and would rather just remove it.

But if the data is used I can undo that

File: lib/shared/src/prompt/truncation.ts (hunk 1 above)
Comment (RXminuS) [reply lib/shared/src/prompt/truncation.ts:21]: It's subtle...but if you look I pass in the TokenCounterUtils from outside so there is no synchronous loading of the library. Instead the loading should be done (async) by whomever calls this function.

But more broadly I've been on the fence if telemetry should be async instead and un-awaited by default. This ensures that it at least never blocks the main functionality. However it might cause issues when doing things like adding events to traces etc, which we also do in several places. By making them sync it's also easier in tests to know that all telemetry events SHOULD have fired.

Comment (akalia25) [reply lib/shared/src/prompt/truncation.ts]: Thanks for confirming, I overlooked the detail of `TokenCounterUtils`. Changing the behavior of telemetry (async, vs sync) would be a larger discussion, lets table that for now :)

**Comment by RXminuS** [comment]
@akalia25 what do you want to do with those failing telemetry events in the other tests? 

Can we somehow just remove the event from the general checks as we can do more detailed testing now? Or do I have to go in and update them all?

File: vscode/src/chat/chat-view/ChatController.ts [hunk 4]
```diff
@@ -622,20 +618,30 @@ export class ChatController implements vscode.Disposable, vscode.WebviewViewProv
...
-            }
-            await this.recordChatQuestionTelemetryEvent(
-                authStatus,
-                mentions,
-                sharedProperties,
-                inputText
+                repoMetadata,
+                repoIsPublic,
+                traceId: span.spanContext().traceId,
+                promptText: inputText,
```
Comment (akalia25) [inline vscode/src/chat/chat-view/ChatController.ts:635]: For the collection of transcript data, we'd need to do the same `isDotCom()` check that was happening before, is this happening somewhere else in the code now?

The reason for this is two-fold, first prevents the local PG instance from storing this data and increasing db load and second for additional security around this sensitive data. 

```suggestion
promptText:
                    isDotCom(authStatus) &&
                    (await truncatePromptString(inputText, CHAT_INPUT_TOKEN_BUDGET)),
            },
```

File: vscode/e2e/telemetry/at-mention.test.ts-snapshots/responseRecievedEvents.public.snap.json [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+          {
+            "key": "userSpecifiedIntent",
+            "value": 1
+          }
+        ],
+        "privateMetadata": {
+          "gitMetadata": "<Object>",
+          "traceId": "<string>",
+          "userSpecifiedIntent": "auto"
+        },
```
Comment (akalia25) [inline vscode/e2e/telemetry/at-mention.test.ts-snapshots/responseRecievedEvents.public.snap.json:147]: based on the recordings, we are no longer collecting `promptText` on `chat-question/executed`?

File: vscode/src/chat/chat-view/ChatController.ts [hunk 6]
```diff
@@ -719,15 +725,16 @@ export class ChatController implements vscode.Disposable, vscode.WebviewViewProv
...
-                            firstTokenSpan,
-                            inputText,
-                            sharedProperties,
-                            userSpecifiedIntent,
-                            detectedIntent: intent,
-                            detectedIntentScores: intentScores,
-                        })
+                        telemetryEvents['cody.chat-question/executed'].record(
+                            {
+                                ...telemetryProperties,
```
Comment (akalia25) [inline vscode/src/chat/chat-view/ChatController.ts:730]: is this intended, to remove `inputText` here?

**Review by akalia25 — DISMISSED** [review]
Left a few comments regarding the collection of transcripts, and changes introduced to that. After those, I can approve :)

**Comment by akalia25** [comment]
> @akalia25 what do you want to do with those failing telemetry events in the other tests?
> 
> Can we somehow just remove the event from the general checks as we can do more detailed testing now? Or do I have to go in and update them all?

hmm I'd love to have those updated 😅 , those have been critical in us spot-checking events. Happy to assist and update them as well!

**Review by akalia25 — APPROVED** [review]
